### PR TITLE
Fix behavior of Trilinos_CXX_COMPILER_FLAGS/Trilinos updates

### PR DESCRIPTION
**Pull request by ndellingwood**

### CODE REVIEW

**Comment by ndellingwood** [comment]
To supercede #2637 
@jjwilke

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
How is this different from #2367? Is this also experimental? If not, it would be great if one of you could provide some more explanations.

**Comment by ndellingwood** [comment]
@masterleinad via email with @jjwilke this is the branch he is actively testing that restores behavior to Trilinos as used by some (possibly most) downstream apps. The previous PR #2637 can probably be closed, I'll check.

**Comment by jjwilke** [comment]
This PR does a few main things:
* Turns off transitive flags in Trilinos since this can blow up the build system
* Makes sure Kokkos flags appear in `Trilinos_CXX_COMPILER_FLAGS`
* Moves functionality into KokkosConfigCommon.cmake

**Comment by jjwilke** [comment]
@dalg24 @masterleinad 
Review this PR for Trilinos fixes

File: cmake/kokkos_tribits.cmake [hunk 1]
```diff
@@ -313,10 +318,18 @@ ENDFUNCTION()
...
-    KOKKOS_SET_LIBRARY_PROPERTIES(${TRIBITS_LIBRARY_NAME} PLAIN_STYLE)
+    #Tribits has way too much techinical debt and baggage to even
+    #allow PUBLIC target_compile_options to be used. It forces C++ flags on projects
+    #as a giant blob of space-separated strings. We end up with duplicated
+    #flags between the flags implicitly forced on Kokkos-dependent and those Kokkos
+    #has in its public INTERFACE_COMPILE_OPTIONS.
+    #These do NOT get de-deduplicated because Tribits
+    #creates flags as a giant monolithic space-separated string
+    #Do not set any transitive properties and keep everything working as before
+    #KOKKOS_SET_LIBRARY_PROPERTIES(${TRIBITS_LIBRARY_NAME} PLAIN_STYLE PRIVATE)
```
Comment (masterleinad) [inline cmake/kokkos_tribits.cmake]: Do I understand correctly that this line is indeed not used and all the modifications to `KOKKOS_SET_LIBRARY_PROPERTIES` are not necessary at all?

Comment (jjwilke) [reply cmake/kokkos_tribits.cmake]: Yeah, I initially wanted the ability to add flags `PRIVATE`ly, but I can revert that.

Comment (masterleinad) [reply cmake/kokkos_tribits.cmake]: OK. If we don't use it, let's just revert it.

**Comment by ndellingwood** [comment]
@jjwilke is this safe for merge after recent commits?

**Comment by ndellingwood** [comment]
Some failures in CI (ignoring the Travis CI test failures):

The gcc failures I think have occurred sporadically like below:

```
[GCC-4.8.4] unexpected EOF
...
Failed in branch GCC-4.8.4
```


Are the CUDA errors legit or randomly occurring?

```
[CUDA-9.2-NVCC]
27/32 Test #27: PerfTestExec ...........................***Failed  Error regular expression found in output. Regex=[  FAILED  ] 14.35 sec

[ RUN      ] default_exec.overlap_mdrange_policy
Time MDRangePolicy: NonOverlap: 0.007645 Time Overlap: 0.003855
/var/jenkins/workspace/Kokkos/core/perf_test/PerfTest_ExecSpacePartitioning.cpp:462: Failure
Value of: time_overlapped_reduce < 1.5 * time_no_overlapped_reduce
  Actual: false
Expected: true
[  FAILED  ] default_exec.overlap_mdrange_policy (38 ms)
```

**Comment by ndellingwood** [comment]
Retest this please.

**Comment by jjwilke** [comment]
I believe this is safe to merge - but I don't understand the test failure. The changes should have no affect on the Travis/Jenkins testing.

**Comment by ndellingwood** [comment]
@jjwilke yeah, these changes don't impact that test. Glancing at the test that failed I think it is random - it relies on a timing data comparison to determine pass/fail, and the timing on that test will have fluctuation so an outlier in performance time can cause it to fail is my guess.

**Pull request merged**